### ISSUE TITLE
Allow `Gen.Params` to be subclassed

### ISSUE THREAD

**Issue by ieure** [rickynils/scalacheck#51]
For some of the more involved types I'm generating, it would be very nice to subclass `Gen.Params` to control the generator at a fine-grained level. For example, I need to test a case class representing the creation of a bank account, and I need to generate both valid routing/account numbers and invalid / unset ones to ensure that bad requests are detected and handled. Using `suchThat` with more than one field on a case class nearly always results in generator exhaustion, so I need to explicitly generate valid/invalid types.

Currently, I'm doing this with a function like:

``` scala
def accountGenerator(genRoutingAccount: Gen[(Option[String], Option[String]]) = for {(routing, account) <- genRoutingAccount … } yield CreateAccount(…)}
```

But it seems more idiomatic to use `Gen.Params` to control things like this. This is not currently possible because, while I can subclass `Gen.Params` (with a normal class, not a case class):

``` scala
  class CustomParams extends Gen.Params

  val tg: Gen[String] = Gen({p: CustomParams => Option("x")})
```

It fails with:

```
ERROR: type mismatch;  found   : TestGenerator.this.CustomParams => Option[java.lang.String]  required: org.scalacheck.Gen.Params => Option[String] : line 221
```

This is because the type bounds of `Gen.apply` don't accept subclasses of `Gen.Params`.
